src/modulefinder/sentry_modulefinder_apple.c

```diff
@@ -130,19 +130,39 @@ remove_image(const struct mach_header *mh, intptr_t UNUSED(vmaddr_slide))
 sentry_value_t
 sentry_get_modules_list(void)
 {
+    // We have 2 locked blocks here (actually 3, with the one inside of the
+    // `add_image` callback). We do that because we have observed deadlocks when
+    // code concurrently `dlopen`s and thus invokes the `add_image` callback
+    // from a different thread.
+    bool should_initialize = false;
     sentry__mutex_lock(&g_mutex);
-    if (!g_initialized) {
+    if (g_initialized) {
+        // avoid an unlock/lock cycle in this case
+        goto is_initialized;
+    } else {
         g_modules = sentry_value_new_list();
+        g_initialized = true;
+        should_initialize = true;
+    }
+    sentry__mutex_unlock(&g_mutex);
```

Comment:
You could just keep the initial structure of the function and put unlock+lock in the branch you had there. In pseudo-code:

```c
lock();
if (!g_initialized) {
  g_initialized = true;
  g_modules = new_list();

  unlock();
  //expensive stuff here
  lock();
}
incref()
return 
```

Reply:
thats a very good idea, makes things more readable and explicit!